### ISSUE TITLE
Prevent pickling of `Result` objects.

### ISSUE THREAD

**Issue by jakereps** [qiime2/qiime2#296]
This was discovered by a QIIME 2 user after running a loop of diversity methods, and storing the results in a dictionary. In order to prevent having to rerun the methods the dictionaries were simply pickled and saved to disk. Initially this was tested to see if it worked by running it and loading the pickled dictionary into a second variable and comparing the two, which passed as (IIRC) only the uuids and some other internal values are compared.

The thing here is that once the artifact left memory its temporary data path is cleaned up, so when the pickled dictionary was loaded after a Jupyter kernel restart, the framework was throwing errors when trying to interact with the objects, saying the `InPath(...) does not exist` on an attempted `.view()`.

With the way the `Result` object works, it seems that attempted pickling should be caught and prevented, or somehow made to properly reload itself once it is un-pickled.

**Comment by jairideout** [comment]
The ability to pickle `Result` objects is actually intentional -- `Result` objects are required to be pickled when `Actions` are executed with `.async()` (this is a requirement of the `multiprocessing` module used under the hood).

A `Result` object's temporary directory is kept around if it isn't owned by the process attempting to destroy the object. This is how we're able to keep the temporary directory around long enough for a `Result` to be pickled in one process and unpickled in another when executing actions asynchronously.

In the past we discussed the issue that you just ran into, where a pickled `Result` wouldn't be able to be unpickled after its owning process has died. We thought it was probably okay for pickled `Result` objects to be ephemeral since that meant we could use `concurrent.futures`, but otherwise we'd likely disallow pickling of `Result` objects to prevent this kind of confusion.

Would it be helpful to have a detailed error message raised when a `Result` is attempted to be unpickled and its temporary directory no longer exists? That would at least make an error happen sooner, and the error message could give the user some hints about what might have gone wrong. We can set up this behavior by implementing `__getstate__` and `__setstate__` on `Result` (to override default pickle behavior).

**Comment by jakereps** [comment]
Yeah, that could be enough. I remember the headache now of trying to get `.async()` to work properly. Would displaying a warning pre-pickling help users, so that it's not an after-the-fact 'oh, your data is gone, run your process again' type of thing? Then the `.async()` call could just suppress that somehow, like a keyword arg that it's being done internally vs `pickle.dumps(..)`.

**Comment by jairideout** [comment]
That's a great idea! We could have `__getstate__` raise the warning whenever it's called, and `.async` could catch the warning and suppress it (using the `warnings` module to filter the warning). `__setstate__` could also benefit from a cleaner error message when the temp dir has already been cleaned up.

Are you interested in working on this?

**Comment by jakereps** [comment]
Sure, I can take a swing at it!

**Comment by jairideout** [comment]
Sweet, thanks @jakereps! I'm happy to review the PR.

**Comment by jakereps** [comment]
Oof, this one got lost in the mail - gladly would take a look at it though, granted things haven't changed w.r.t. async/Pickling. Will look tomorrow and see if I can knock this out our close it if it's no longer an issues.
